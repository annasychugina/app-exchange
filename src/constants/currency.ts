import {Currency, SymbolCurrency} from '../types/types';

export const CURRENCIES: Array<Currency> = ['RUB', 'USD', 'EUR', 'GBP'];

export const UPDATE_RATES_DELAY = 10000000;

export const CURRENCY_SYMBOL_MAP: {[key in Currency]: SymbolCurrency} = {
  RUB: '₽',
  USD: '$',
  EUR: '€',
  GBP: '£',
};
