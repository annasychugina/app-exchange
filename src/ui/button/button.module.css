.button {
  display: inline-block;
  line-height: 20px;
  font-size: 16px;
  font-weight: 500;
  text-align: center;
  color: white;
  cursor: pointer;
  background-color: rgb(235, 0, 141);
  transition: all 0.2s ease 0s;
  outline: none;
  border-radius: 40px;
  user-select: none;
  padding: 10px 24px;
  border-style: none;

  &:disabled {
    opacity: .2;
    pointer-events: none;
  }

  &:hover {
    background-color: rgb(210, 0, 126);
  }
  &:active {
    background-color: rgb(235, 0, 141);
  }
}

.loader {
  margin-right: auto;
  margin-left: auto;
  display: block;
  animation: spinner-rotate 3s linear infinite;
}

.circle {
  fill: none;
  stroke-dasharray: 280;
  stroke-dashoffset: 100;
  stroke-linecap: round;
  stroke-width: 5;
  animation: spinner-offset 3s linear infinite;
  stroke: white;
}


@keyframes spinner-rotate {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(1080deg);
  }
}
