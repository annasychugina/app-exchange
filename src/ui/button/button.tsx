import React from 'react';
import cn from 'classnames';

import s from './button.module.css';

type Props = {
  title?: string;
  disabled?: boolean;
  className?: string;
  children: any;
  onClick: () => void;
  hasLoader?: boolean;
  type?: string;
};

export const Button: React.FC<Props> = ({
  type = 'button',
  hasLoader = false,
  disabled = false,
  className,
  title,
  onClick,
  children,
}: Props): React.ReactElement => (
  <button className={cn(s.button, className)} disabled={disabled} onClick={onClick} title={title} aria-label={title}>
    {hasLoader ? (
      <svg width={15} height={15} version="1.1" x="0" y="0" className={s.loader} viewBox="0 0 100 100">
        <circle className={s.circle} cx="50" cy="50" r="44" />
      </svg>
    ) : (
      children
    )}
  </button>
);
