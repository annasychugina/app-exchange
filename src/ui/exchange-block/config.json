{
  "ExchangeBlock": {
    "balanceText": "You have %balance% %currency%"
  }
}
