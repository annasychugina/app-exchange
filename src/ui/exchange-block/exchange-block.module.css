.root {
  &_currencyFrom {
    position: relative;
   background: linear-gradient(to bottom, #1E91E4, #3F90EA);
    &::after {
      content: '';
      display: block;
      position: absolute;
      transform: translate(-50%, -50%) rotate(45deg);
      width: 20px;
      height: 20px;
      left: 50%;
      top: 100%;
      background-color: #3F90EA;
    }
  }

  &_currencyTo {
    background: linear-gradient(to bottom,  #2351C2, #2757CC);
  }
}

.row {
  display: flex;
  width: 100%;
  justify-content: space-between;
  align-items: center;
  &_rate {
    justify-content: flex-end;
  }
}

.block {
  box-sizing: border-box;
  padding: 40px 50px 46px;
  display: flex;
  flex-wrap: wrap;
  font-size: 26px;
  color: white;
}

.text {
  margin-top: 15px;
  color: rgba(255,255,255, .85);
  font-size: 12px;
}
