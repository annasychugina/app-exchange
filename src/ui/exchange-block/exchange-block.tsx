import React from 'react';
import cn from 'classnames';
import {Currency, CurrencyBlockType, FieldInputCallback} from '../../types/types';
import {NumberFormatInput} from '../number-format-input/number-format-input';
import {Slider} from '../slider/slider';
import {CURRENCY_SYMBOL_MAP} from '../../constants/currency';
import {resourcesTemplate} from '../../utils/resourcesTemplate';
import {formattedValue} from '../../utils/formatters';
import {parseMoney} from '../../utils/parseMoney';
import resources from './config.json';

import s from './exchange-block.module.css';

interface Props {
  currency: Currency;
  rate?: number;
  inputValue: string;
  type: CurrencyBlockType;
  onCurrencyChange: (currency: Currency) => void;
  onCurrencyValueChange: FieldInputCallback;
  currencyItems: Array<Currency>;
  balance: number;
  currencyAmountStr: string;
  isRateLoading?: boolean;
}

export const ExchangeBlock: React.FC<Props> = ({
  currencyItems,
  rate,
  currency,
  onCurrencyChange,
  onCurrencyValueChange,
  type,
  inputValue,
  balance,
  currencyAmountStr,
}) => {
  const handleSlide = (index: number): void => {
    onCurrencyChange(currencyItems[index]);
  };
  return (
    <div className={cn(s.root, s[`root_${type}`])}>
      <Slider onSlideChange={handleSlide} currentSlide={currencyItems.indexOf(currency)}>
        {currencyItems.map((currency, index) => (
          <div key={currency} className={s.block}>
            <div className={s.row}>
              <p>{currency}</p>
              <NumberFormatInput
                prefix={type === CurrencyBlockType.currencyFrom ? '-' : '+'}
                name={type}
                inputValue={inputValue}
                onChange={onCurrencyValueChange}
              />
            </div>

            <div className={s.row}>
              <p className={s.text}>
                {' '}
                {resourcesTemplate(resources.ExchangeBlock.balanceText, {
                  balance: parseMoney(currency, formattedValue(balance))
                })}
              </p>
            </div>
            {rate && type === CurrencyBlockType.currencyTo && (
              <div className={cn(s.row, s.row_rate)}>
                <p className={cn(s.text, s.text_rate)}>{currencyAmountStr}</p>
              </div>
            )}
          </div>
        ))}
      </Slider>
    </div>
  );
};
