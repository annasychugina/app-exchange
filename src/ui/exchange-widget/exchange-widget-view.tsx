import React from 'react';
import {Currency, CurrencyBlockType, FormState} from '../../types/types';
import {FieldInputCallback} from '../formatted-input/formatted-input';
import {ExchangeBlock} from '../exchange-block/exchange-block';
import {CURRENCIES} from '../../constants/currency';
import {Button} from '../button/button';
import resources from './config.json';

import s from './exchange-widget-view.module.css';

interface Props {
  currencyFrom: Currency;
  currencyTo: Currency;
  handleCurrencyFromChange: (currency: Currency) => void;
  handleCurrencyToChange: (currency: Currency) => void;
  handleCurrencyValueChange: FieldInputCallback;
  rate: number;
  isRateLoading: boolean;
  valueFrom?: FormState['currencyFrom'];
  valueTo?: FormState['currencyFrom'];
}

export const ExchangeWidgetView: React.FC<Props> = ({
  currencyTo,
  currencyFrom,
  handleCurrencyFromChange,
  handleCurrencyToChange,
  handleCurrencyValueChange,
  rate,
  valueFrom,
  valueTo,
  isRateLoading,
}) => {
  return (
    <section className={s.root}>
      <div className={s.content}>
        <ExchangeBlock
          onCurrencyChange={currency => {
            handleCurrencyFromChange(currency);
          }}
          currencyTo={currencyTo}
          currencyFrom={currencyFrom}
          type={CurrencyBlockType.currencyFrom}
          currencyItems={CURRENCIES}
          onCurrencyValueChange={handleCurrencyValueChange}
          inputValue={String(valueFrom)}
        />
        <ExchangeBlock
          rate={rate}
          onCurrencyChange={currency => {
            handleCurrencyToChange(currency);
          }}
          currencyTo={currencyTo}
          currencyFrom={currencyFrom}
          type={CurrencyBlockType.currencyTo}
          currencyItems={CURRENCIES}
          onCurrencyValueChange={handleCurrencyValueChange}
          inputValue={String(valueTo)}
        />
      </div>
      <Button
        type="submit"
        className={s.submitBtn}
        onClick={() => console.log('click')}
        title={resources.ExchangeWidget.buttonTitle}
      >
        {resources.ExchangeWidget.buttonText}
      </Button>
    </section>
  );
};
