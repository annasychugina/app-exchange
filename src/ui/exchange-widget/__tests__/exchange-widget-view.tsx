import React from 'react';
import {CURRENCIES} from '../../../constants/currency';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {ExchangeWidgetView} from '../exchange-widget-view';
import {ExchangeBlock} from '../../exchange-block/exchange-block';
import {Button} from '../../button/button';

const noop = () => {};
const mockProps = {
  currencyFrom: CURRENCIES[0],
  currencyTo: CURRENCIES[1],
  balanceFrom: 230,
  balanceTo: 120,
  handleCurrencyFromChange: noop,
  handleCurrencyToChange: noop,
  handleCurrencyValueChange: noop,
  handleExchange: noop,
  rate: 1.2,
  isRateLoading: false,
  valueFrom: 12,
  valueTo: 10,
  isExchangeButtonDisabled: true,
};

it('renders correctly', () => {
  const wrapper = shallow(<ExchangeWidgetView {...mockProps} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('calls onCurrencyValueChange first ExchangeBlock', () => {
  const onCurrencyValueChange = jest.fn();
  const wrapper = shallow(<ExchangeWidgetView {...mockProps} handleCurrencyValueChange={onCurrencyValueChange} />);

  const rows = wrapper.find(ExchangeBlock);
  expect(rows).toHaveLength(2);

  rows
    .at(0)
    .props()
    .onCurrencyValueChange(35, 'currencyFrom');

  expect(onCurrencyValueChange).toBeCalledTimes(1);
  expect(onCurrencyValueChange).toHaveBeenLastCalledWith(35, 'currencyFrom');
});

it('calls onCurrencyValueChange first ExchangeBlock', () => {
  const onCurrencyValueChange = jest.fn();
  const wrapper = shallow(<ExchangeWidgetView {...mockProps} handleCurrencyValueChange={onCurrencyValueChange} />);
  const blocks = wrapper.find(ExchangeBlock);
  expect(blocks).toHaveLength(2);
  blocks
    .at(1)
    .props()
    .onCurrencyValueChange(35, 'currencyFrom');
  expect(onCurrencyValueChange).toBeCalledTimes(1);
  expect(onCurrencyValueChange).toHaveBeenLastCalledWith(35, 'currencyFrom');
});

it('disable button if prop isExchangeButtonDisabled', () => {
  const wrapper1 = shallow(<ExchangeWidgetView {...mockProps} isExchangeButtonDisabled={true} />);
  const button = wrapper1.find(Button);
  expect(button).toHaveLength(1);
  expect(button.props().disabled).toBe(true);
  const wrapper2 = shallow(<ExchangeWidgetView {...mockProps} isExchangeButtonDisabled={false} />);
  const button2 = wrapper2.find(Button);
  expect(button2).toHaveLength(1);
  expect(button2.props().disabled).toBe(false);
});

it('calls handleExchange when click button', () => {
  const onExchange = jest.fn();
  const e = jest.fn();
  const wrapper = shallow(
    <ExchangeWidgetView {...mockProps} isExchangeButtonDisabled={false} handleExchange={onExchange} />,
  );
  const button = wrapper.find(Button);
  expect(button).toHaveLength(1);
  expect(button.props().onClick).toBeTruthy();
  button.props().onClick!(e as any);
  expect(onExchange).toBeCalledTimes(1);
});

it('calls onCurrencyChange first ExchangeBlock', () => {
  const onCurrencyFromChange = jest.fn();
  const wrapper = shallow(<ExchangeWidgetView {...mockProps} handleCurrencyFromChange={onCurrencyFromChange} />);
  const blocks = wrapper.find(ExchangeBlock);
  blocks
    .at(0)
    .props()
    .onCurrencyChange('RUB');
  expect(onCurrencyFromChange).toBeCalledTimes(1);
  expect(onCurrencyFromChange).toHaveBeenLastCalledWith('RUB');
});

it('calls onCurrencyChange second ExchangeBlock', () => {
  const onCurrencyToChange = jest.fn();
  const wrapper = shallow(<ExchangeWidgetView {...mockProps} handleCurrencyToChange={onCurrencyToChange} />);
  const blocks = wrapper.find(ExchangeBlock);
  blocks
    .at(1)
    .props()
    .onCurrencyChange('RUB');
  expect(onCurrencyToChange).toBeCalledTimes(1);
  expect(onCurrencyToChange).toHaveBeenLastCalledWith('RUB');
});
