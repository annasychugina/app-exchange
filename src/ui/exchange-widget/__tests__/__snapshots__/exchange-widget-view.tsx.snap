// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="root"
>
  <div
    className="content"
  >
    <ExchangeBlock
      balance={230}
      currency="RUB"
      currencyAmountStr="$1 = ₽1.2"
      currencyItems={
        Array [
          "RUB",
          "USD",
          "EUR",
          "GBP",
        ]
      }
      inputValue="12"
      onCurrencyChange={[Function]}
      onCurrencyValueChange={[Function]}
      type="currencyFrom"
    />
    <ExchangeBlock
      balance={120}
      currency="USD"
      currencyAmountStr="$1 = ₽1.2"
      currencyItems={
        Array [
          "RUB",
          "USD",
          "EUR",
          "GBP",
        ]
      }
      inputValue="10"
      isRateLoading={false}
      onCurrencyChange={[Function]}
      onCurrencyValueChange={[Function]}
      rate={1.2}
      type="currencyTo"
    />
  </div>
  <Button
    className="submitBtn"
    disabled={true}
    onClick={[Function]}
    title="Exchange"
  >
    Exchange money
  </Button>
</section>
`;
